ledger/src/shred/merkle.rs

```diff
@@ -1178,219 +1244,191 @@ pub(super) fn make_shreds_from_data(
                 ))
             })
             .ok_or(Error::UnknownProofSize)?;
-        common_header.shred_variant = ShredVariant::MerkleData {
+        common_header_data.shred_variant = ShredVariant::MerkleData {
             proof_size,
             chained,
             resigned,
         };
-        common_header.fec_set_index = common_header.index;
-        for shred in data
-            .chunks(data_buffer_size)
-            .chain(std::iter::repeat(&[][..]))
-            .take(num_data_shreds)
-        {
-            let shred = new_shred_data(common_header, data_header, shred);
-            shreds.push(shred);
-            common_header.index += 1;
-        }
-        if let Some(shred) = shreds.last() {
+        common_header_code.shred_variant = ShredVariant::MerkleCode {
+            proof_size,
+            chained,
+            resigned,
+        };
+        common_header_data.fec_set_index = common_header_data.index;
+        common_header_code.fec_set_index = common_header_data.fec_set_index;
+        shreds.extend({
+            let chunks = data
+                .chunks(data_buffer_size)
+                .chain(std::iter::repeat(&[][..])) // possible padding
+                .take(num_data_shreds);
+            make_shreds_data(&mut common_header_data, data_header, chunks).map(Shred::ShredData)
+        });
+        if let Some(Shred::ShredData(shred)) = shreds.last() {
```

Comment:
For my understanding, at this point there must be at least one data shred here right? Is it better to `expect`?

Reply:
I believe that is true.
Though it is a fair sanity check, I am a bit hesitant to add new panics into the code, in case some subtle condition is missed.
I will add a `debug_assert` though in a follow up PR.